File: app-android/src/main/java/com/ustadmobile/port/android/view/ContentEntryDetailActivity.java
package com.ustadmobile.port.android.view;

import android.Manifest;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.flexbox.FlexDirection;
import com.google.android.flexbox.FlexboxLayoutManager;
import com.squareup.picasso.Picasso;
import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.ContentEntryDetailPresenter;
import com.ustadmobile.core.generated.locale.MessageID;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.networkmanager.LocalAvailabilityListener;
import com.ustadmobile.core.networkmanager.LocalAvailabilityMonitor;
import com.ustadmobile.core.util.UMFileUtil;
import com.ustadmobile.core.view.ContentEntryDetailView;
import com.ustadmobile.lib.db.entities.ContentEntryRelatedEntryJoinWithLanguage;
import com.ustadmobile.port.android.netwokmanager.NetworkManagerAndroidBle;
import com.ustadmobile.port.android.util.UMAndroidUtil;
import com.ustadmobile.port.sharedse.networkmanager.NetworkManagerBle;

import java.util.HashMap;
import java.util.Hashtable;
import java.util.List;
import java.util.Set;

import static com.ustadmobile.core.controller.ContentEntryDetailPresenter.LOCALLY_AVAILABLE_ICON;
import static com.ustadmobile.core.controller.ContentEntryDetailPresenter.LOCALLY_NOT_AVAILABLE_ICON;

public class ContentEntryDetailActivity  extends UstadBaseActivity implements
        ContentEntryDetailView, ContentEntryDetailLanguageAdapter.AdapterViewListener,
        LocalAvailabilityMonitor, LocalAvailabilityListener {

    private ContentEntryDetailPresenter entryDetailPresenter;

    private NetworkManagerAndroidBle managerAndroidBle;

    private TextView localAvailabilityStatusText;

    private TextView entryDetailsTitle;

    private TextView entryDetailsDesc;

    private TextView entryDetailsLicense;

    private TextView entryDetailsAuthor;

    private TextView translationAvailableLabel;

    private TextView downloadSize;

    private ImageView localAvailabilityStatusIcon;

    private RecyclerView flexBox;

    private Button downloadButton;

    private DownloadProgressView downloadProgress;

    HashMap<Integer,Integer> fileStatusIcon = new HashMap<>();


    @Override
    protected void onBleNetworkServiceBound(NetworkManagerBle networkManagerBle) {
        super.onBleNetworkServiceBound(networkManagerBle);
        managerAndroidBle = (NetworkManagerAndroidBle) networkManagerBle;
        entryDetailPresenter = new ContentEntryDetailPresenter(getContext(),
                UMAndroidUtil.bundleToHashtable(getIntent().getExtras()), this,this);
        entryDetailPresenter.onCreate(UMAndroidUtil.bundleToHashtable(new Bundle()));
        entryDetailPresenter.onStart();
        managerAndroidBle.addLocalAvailabilityListener(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_entry_detail);

        localAvailabilityStatusText = findViewById(R.id.content_status_text);
        localAvailabilityStatusIcon = findViewById(R.id.content_status_icon);
        downloadButton = findViewById(R.id.entry_detail_button);
        downloadProgress = findViewById(R.id.entry_detail_progress);
        entryDetailsTitle = findViewById(R.id.entry_detail_title);
        entryDetailsDesc = findViewById(R.id.entry_detail_description);
        entryDetailsLicense = findViewById(R.id.entry_detail_license);
        entryDetailsAuthor = findViewById(R.id.entry_detail_author);
        downloadSize = findViewById(R.id.entry_detail_content_size);
        translationAvailableLabel = findViewById(R.id.entry_detail_available_label);
        flexBox = findViewById(R.id.entry_detail_flex);

        fileStatusIcon.put(LOCALLY_AVAILABLE_ICON,R.drawable.ic_nearby_black_24px);
        fileStatusIcon.put(LOCALLY_NOT_AVAILABLE_ICON,R.drawable.ic_cloud_download_black_24dp);

        setUMToolbar(R.id.entry_detail_toolbar);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Respond to the action bar's Up/Home button
            case android.R.id.home:
                clickUpNavigation();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void clickUpNavigation() {
        runOnUiThread(() -> {
            if (entryDetailPresenter != null) {
                entryDetailPresenter.handleUpNavigation();
            }
        });


    }


    @Override
    public void setContentEntryTitle(String title) {
        entryDetailsTitle.setText(title);
        getSupportActionBar().setTitle(title);
    }

    @Override
    public void setContentEntryDesc(String desc) {
        entryDetailsDesc.setText(desc);
    }

    @Override
    public void setContentEntryLicense(String license) {
        entryDetailsLicense.setText(license);
    }

    @Override
    public void setContentEntryAuthor(String author) {
        entryDetailsAuthor.setText(author);
    }

    @Override
    public void setDetailsButtonEnabled(boolean enabled) {
        downloadButton.setEnabled(enabled);
    }

    @Override
    public void setDownloadSize(long fileSize) {
        downloadSize.setText(UMFileUtil.formatFileSize(fileSize));
    }

    @Override
    public void loadEntryDetailsThumbnail(String thumbnailUrl) {

        Picasso.with(ContentEntryDetailActivity.this)
                .load(thumbnailUrl)
                .into((ImageView) findViewById(R.id.entry_detail_thumbnail));
    }

    @Override
    public void setAvailableTranslations(List<ContentEntryRelatedEntryJoinWithLanguage> result,
                                         long entryUuid) {

        FlexboxLayoutManager flexboxLayoutManager = new FlexboxLayoutManager(getApplicationContext());
        flexboxLayoutManager.setFlexDirection(FlexDirection.ROW);
        flexBox.setLayoutManager(flexboxLayoutManager);

        ContentEntryDetailLanguageAdapter adapter = new ContentEntryDetailLanguageAdapter(result,
                this, entryUuid);
        flexBox.setAdapter(adapter);

    }

    @Override
    public void updateDownloadProgress(float progressValue) {
        downloadProgress.setProgress(progressValue);
    }

    @Override
    public void setDownloadButtonVisible(boolean visible) {
        downloadButton.setVisibility(visible ? View.VISIBLE : View.GONE);

    }



    @Override
    public void setButtonTextLabel(String textLabel) {
        downloadButton.setText(textLabel);
    }

    @Override
    public void showFileOpenError() {
        Toast.makeText((Context) getContext(), R.string.error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void updateLocalAvailabilityViews(int icon, String status) {
        localAvailabilityStatusText.setVisibility(View.VISIBLE);
        localAvailabilityStatusIcon.setVisibility(View.VISIBLE);
        localAvailabilityStatusIcon.setImageResource(fileStatusIcon.get(icon));
        localAvailabilityStatusText.setText(status);
    }

    @Override
    public void setLocalAvailabilityStatusViewVisible(boolean visible) {
        localAvailabilityStatusIcon.setVisibility(visible ? View.VISIBLE : View.GONE);
        localAvailabilityStatusText.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setTranslationLabelVisible(boolean visible) {
        translationAvailableLabel.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setFlexBoxVisible(boolean visible) {
        flexBox.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setDownloadProgressVisible(boolean visible) {
        downloadProgress.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setDownloadProgressLabel(String progressLabel) {
        downloadProgress.setStatusText(progressLabel);
    }

    @Override
    public void setDownloadButtonClickableListener(boolean isDownloadComplete) {
        downloadButton.setOnClickListener(view ->
                entryDetailPresenter.handleDownloadButtonClick(isDownloadComplete,
                        entryDetailPresenter.getEntryUuid()));
    }

    @Override
    public void showDownloadOptionsDialog(Hashtable args) {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        runAfterGrantingPermission(
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                () -> impl.go("DownloadDialog", args, getContext()),
                impl.getString(MessageID.download_storage_permission_title,getContext()),
                impl.getString(MessageID.download_storage_permission_message,getContext()));
    }

    @Override
    public void selectContentEntryOfLanguage(long uid) {
        entryDetailPresenter.handleClickTranslatedEntry(uid);
    }

    @Override
    public void startMonitoringAvailability(Object monitor, List<Long> entryUidsToMonitor) {
        managerAndroidBle.startMonitoringAvailability(monitor,entryUidsToMonitor);
    }

    @Override
    public void stopMonitoringAvailability(Object monitor) {
        managerAndroidBle.stopMonitoringAvailability(monitor);
    }

    @Override
    public void onDestroy() {
        entryDetailPresenter.onDestroy();
        networkManagerBle.removeLocalAvailabilityListener(this);
        super.onDestroy();
    }

    @Override
    public void onLocalAvailabilityChanged(Set<Long> locallyAvailableEntries) {
        entryDetailPresenter.handleLocalAvailabilityStatus(locallyAvailableEntries);
    }
}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/ContentEntryDetailActivity.kt
package com.ustadmobile.port.android.view

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Html
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.core.widget.NestedScrollView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.ContentEntryDetailPresenter
import com.ustadmobile.core.controller.ContentEntryDetailPresenter.Companion.LOCALLY_AVAILABLE_ICON
import com.ustadmobile.core.controller.ContentEntryDetailPresenter.Companion.LOCALLY_NOT_AVAILABLE_ICON
import com.ustadmobile.core.db.JobStatus
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.generated.locale.MessageID
import com.ustadmobile.core.impl.AppConfig
import com.ustadmobile.core.impl.UMAndroidUtil
import com.ustadmobile.core.impl.UMAndroidUtil.bundleToMap
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.util.UMFileUtil
import com.ustadmobile.core.util.mimeTypeToPlayStoreIdMap
import com.ustadmobile.core.view.ContentEntryDetailView
import com.ustadmobile.lib.db.entities.ContentEntry
import com.ustadmobile.lib.db.entities.ContentEntryRelatedEntryJoinWithLanguage
import com.ustadmobile.lib.db.entities.DownloadJobItem
import com.ustadmobile.port.android.view.ext.makeSnackbarIfRequired
import com.ustadmobile.sharedse.network.NetworkManagerBle
import kotlinx.android.synthetic.main.activity_entry_detail.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch


class ContentEntryDetailActivity : UstadBaseWithContentOptionsActivity(),
        ContentEntryDetailView, ContentEntryDetailLanguageAdapter.AdapterViewListener,
        DownloadProgressView.OnStopDownloadListener {

    private lateinit var presenter: ContentEntryDetailPresenter

    private lateinit var managerAndroidBle: NetworkManagerBle

    private var localAvailabilityStatusText: TextView? = null

    private var entryDetailsTitle: TextView? = null

    private var entryDetailsDesc: TextView? = null

    private var entryDetailsLicense: TextView? = null

    private var entryDetailsAuthor: TextView? = null

    private var translationAvailableLabel: TextView? = null

    private var downloadSize: TextView? = null

    private var localAvailabilityStatusIcon: ImageView? = null

    private lateinit var editButton: FloatingActionButton

    private lateinit var flexBox: RecyclerView

    private lateinit var downloadButton: Button

    private lateinit var umAppRepository: UmAppDatabase

    private var downloadProgress: DownloadProgressView? = null

    private var fileStatusIcon = HashMap<Int, Int>()

    private var showControls : Boolean = false

    private var showExportIcon: Boolean = false

    private var currentDownloadJobItemStatus: Int = -1

    override fun onBleNetworkServiceBound(networkManagerBle: NetworkManagerBle) {
        super.onBleNetworkServiceBound(networkManagerBle)
        if (networkManagerBle.isVersionKitKatOrBelow) {
            downloadButton.setBackgroundResource(
                    R.drawable.pre_lollipop_btn_selector_bg_entry_details)
        }

        managerAndroidBle = networkManagerBle
        presenter = ContentEntryDetailPresenter(this,
                bundleToMap(intent.extras), this, true,
                umAppRepository, UmAppDatabase.getInstance(baseContext),
                networkManagerBle.localAvailabilityManager,
                networkManagerBle.containerDownloadManager,
                UmAccountManager.getActiveAccount(viewContext),
                UstadMobileSystemImpl.instance)
        presenter.handleShowEditControls(showControls)
        presenter.onCreate(bundleToMap(Bundle()))

        managerAndroidBle.enablePromptsSnackbarManager.makeSnackbarIfRequired(
                findViewById(R.id.coordinationLayout), this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        umAppRepository = UmAccountManager.getRepositoryForActiveAccount(this)
        showControls = UstadMobileSystemImpl.instance.getAppConfigString(
                AppConfig.KEY_SHOW_CONTENT_EDITOR_CONTROLS, "false", this)!!.toBoolean()
        setContentView(R.layout.activity_entry_detail)

        localAvailabilityStatusText = findViewById(R.id.content_status_text)
        localAvailabilityStatusIcon = findViewById(R.id.content_status_icon)
        downloadButton = findViewById(R.id.entry_download_open_button)
        downloadProgress = findViewById(R.id.entry_detail_progress)
        entryDetailsTitle = findViewById(R.id.entry_detail_title)
        entryDetailsDesc = findViewById(R.id.entry_detail_description)
        entryDetailsLicense = findViewById(R.id.entry_detail_license)
        entryDetailsAuthor = findViewById(R.id.entry_detail_author)
        downloadSize = findViewById(R.id.entry_detail_content_size)
        translationAvailableLabel = findViewById(R.id.entry_detail_available_label)
        flexBox = findViewById(R.id.entry_detail_flex)
        coordinatorLayout = findViewById(R.id.coordinationLayout)
        editButton = findViewById(R.id.edit_content)

        fileStatusIcon[LOCALLY_AVAILABLE_ICON] = R.drawable.ic_nearby_black_24px
        fileStatusIcon[LOCALLY_NOT_AVAILABLE_ICON] = R.drawable.ic_cloud_download_black_24dp

        setUMToolbar(R.id.um_toolbar)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        setProgressBar()
        showBaseProgressBar(false)

        downloadProgress!!.setOnStopDownloadListener(this)

        findViewById<NestedScrollView>(R.id.nested_scroll).setOnScrollChangeListener { v: NestedScrollView?, scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int ->
            if(showControls){
                if (scrollY > oldScrollY) {
                    editButton.hide()
                } else {
                    editButton.show()
                }
            }
        }

        editButton.setOnClickListener {
            presenter.handleStartEditingContent()
        }
        downloadButton.setOnClickListener {
            presenter.handleDownloadButtonClick()
        }
    }

    override fun onResume() {
        super.onResume()
        if(::managerAndroidBle.isInitialized) {
            managerAndroidBle.enablePromptsSnackbarManager.makeSnackbarIfRequired(
                    findViewById(R.id.coordinationLayout), this)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_content_entry_details, menu)
        return super.onCreateOptionsMenu(menu)
    }


    @SuppressLint("RestrictedApi")
    override fun setEditButtonVisible(show: Boolean) {
       if(::editButton.isInitialized){
           editButton.visibility = if(show) View.VISIBLE else View.GONE
       }
    }

    override fun setExportContentIconVisible(visible: Boolean){
        this.showExportIcon = visible
        invalidateOptionsMenu()
    }


    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        menu!!.findItem(R.id.export_content).isVisible = showExportIcon
        return super.onPrepareOptionsMenu(menu)
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            // Respond to the action bar's Up/Home button
            android.R.id.home -> {
                clickUpNavigation()
                return true
            }

            R.id.export_content ->{
                presenter.handleContentEntryExport()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun clickUpNavigation() {
        runOnUiThread {
            presenter.handleUpNavigation()
        }
    }

    override fun setContentEntry(contentEntry: ContentEntry) {
        entryDetailsTitle!!.text = contentEntry.title
        supportActionBar!!.title = contentEntry.title
        entryDetailsDesc!!.text = if(!contentEntry.description.isNullOrBlank()) Html.fromHtml(contentEntry.description) else ""
        entryDetailsAuthor!!.text = contentEntry.author

        UMAndroidUtil.loadImage(contentEntry.thumbnailUrl,R.drawable.img_placeholder,
                findViewById<View>(R.id.entry_detail_thumbnail) as ImageView)
    }

    override fun setDownloadJobItemStatus(downloadJobItem: DownloadJobItem?) {
        if(currentDownloadJobItemStatus != downloadJobItem?.djiStatus) {
            when {
                //TODO: change this to allow failed etc. probably best done using an extension method in core
                downloadJobItem == null -> {
                    entry_download_open_button.text = resources.getText(R.string.download)
                    entry_download_open_button.visibility = View.VISIBLE
                    entry_detail_progress.visibility = View.GONE
                }

                downloadJobItem.djiStatus == JobStatus.COMPLETE -> {
                    entry_download_open_button.visibility = View.VISIBLE
                    entry_download_open_button.text = resources.getText(R.string.open)
                    entry_detail_progress.visibility = View.GONE
                }

                downloadJobItem.djiStatus < JobStatus.COMPLETE_MIN -> {
                    entry_download_open_button.visibility = View.GONE
                    entry_detail_progress.visibility = View.VISIBLE
                }
            }

            currentDownloadJobItemStatus = downloadJobItem?.djiStatus ?: 0
        }

        if(downloadJobItem != null && downloadJobItem.djiStatus > JobStatus.WAITING_MIN
                && downloadJobItem.djiStatus <= JobStatus.COMPLETE_MIN) {
            entry_detail_progress.progress = if(downloadJobItem.downloadLength > 0) {
                (downloadJobItem.downloadedSoFar.toFloat()) / (downloadJobItem.downloadLength.toFloat())
            }else {
                0f
            }
        }
    }

    override fun setContentEntryLicense(license: String) {
        entryDetailsLicense!!.text = license
    }

    override fun setMainButtonEnabled(enabled: Boolean) {
        downloadButton.isEnabled = enabled
    }

    override fun setDownloadSize(fileSize: Long) {
        downloadSize!!.text = UMFileUtil.formatFileSize(fileSize)
    }

    override fun showFileOpenError(message: String, actionMessageId: Int, mimeType: String) {
        showErrorNotification(message, {
            var appPackageName = mimeTypeToPlayStoreIdMap[mimeType]
            if (appPackageName == null) {
                appPackageName = "cn.wps.moffice_eng"
            }
            try {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$appPackageName")))
            } catch (anfe: android.content.ActivityNotFoundException) {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=$appPackageName")))
            }
        }, actionMessageId)
    }

    override fun showFileOpenError(message: String) {
        showErrorNotification(message, {}, 0)
    }


    override fun updateLocalAvailabilityViews(icon: Int, status: String) {
        localAvailabilityStatusText!!.visibility = View.VISIBLE
        localAvailabilityStatusIcon!!.visibility = View.VISIBLE
        localAvailabilityStatusIcon!!.setImageResource(fileStatusIcon[icon]!!)
        localAvailabilityStatusText!!.text = status
    }

    override fun setLocalAvailabilityStatusViewVisible(visible: Boolean) {
        localAvailabilityStatusIcon!!.visibility = if (visible) View.VISIBLE else View.GONE
        localAvailabilityStatusText!!.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun showDownloadOptionsDialog(map: Map<String, String>) {
        val impl = UstadMobileSystemImpl.instance
        runAfterGrantingPermission(
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                Runnable { impl.go("DownloadDialog", map, this) },
                impl.getString(MessageID.download_storage_permission_title, this),
                impl.getString(MessageID.download_storage_permission_message, this))
    }

    override fun selectContentEntryOfLanguage(contentEntryUid: Long) {
        presenter.handleClickTranslatedEntry(contentEntryUid)
    }

    override fun onDestroy() {
        presenter.onDestroy()
        super.onDestroy()
    }

    override fun setAvailableTranslations(result: List<ContentEntryRelatedEntryJoinWithLanguage>) {

        translationAvailableLabel!!.visibility = if (result.isNotEmpty()) View.VISIBLE else View.GONE
        flexBox.visibility = if (result.isNotEmpty()) View.VISIBLE else View.GONE

        val flexboxLayoutManager = FlexboxLayoutManager(applicationContext)
        flexboxLayoutManager.flexDirection = FlexDirection.ROW
        flexBox.layoutManager = flexboxLayoutManager
        val adapter = ContentEntryDetailLanguageAdapter(result, this)
        flexBox.adapter = adapter
    }

    override fun onClickStopDownload(view: DownloadProgressView) {
        GlobalScope.launch {
            presenter.handleCancelDownload()
        }
    }
}
